### PR TITLE
Fix HID controller output on Windows with common-hid-packet-parser.js

### PR DESCRIPTION
**Pull request by xeruf**
> Specifying the report ID is required on Windows as discovered in [this forum thread](https://mixxx.org/forums/viewtopic.php?f=3&t=11946).

> The problem is that the JS is relying on [an implementation quirk of hidapi on Linux](https://mixxx.org/forums/viewtopic.php?p=38832#p38832).

Migrated from https://github.com/mixxxdj/mixxx/pull/1719, since I actually have a HID controller to test this on. But still not working. At least now nothing is broken that wasn't broken before ^^

Zulip conversation: https://mixxx.zulipchat.com/#narrow/stream/113295-controller-mapping/topic/HID.20Mapping

### CODE REVIEW

**Review by Be-ing — CHANGES_REQUESTED** [review]
Thanks for working on this!

File: res/controllers/common-hid-packet-parser.js [hunk 1]
```diff
@@ -1033,14 +1035,16 @@ HIDController.prototype.parsePacket = function(data,length) {
...
-                packet=undefined;
-                break;
+        if(packet.header !== undefined) {
+            for (var header_byte = 0; header_byte < packet.header.length; header_byte++) {
+                if (packet.header[header_byte] != data[header_byte]) {
+                    packet = undefined;
+                    break;
+                }
             }
+            if (packet === undefined)
```
[the remark this answers is not in the document]
Comment (Xerus2000) [reply res/controllers/common-hid-packet-parser.js]: because it is only set undefined within the for loop. If it had been undefined before, there would have been an error already ^^

File: copy-controllers.sh [hunk 2]
```diff
@@ -0,0 +1 @@
+rm -r $APPDATA/../Local/Mixxx/controllers; cp -vr "res/controllers" "C:/Program Files/Mixxx"
```
[the remark this answers is not in the document]
Comment (Xerus2000) [reply copy-controllers.sh]: That was intentional. I will remove it when I'm done. See my comment above. I work on multiple devices, so i need to commit these things.

File: res/controllers/common-hid-packet-parser.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Be-ing) [reply res/controllers/common-hid-packet-parser.js]: Okay, that makes sense... although the library should not be assigning variables to `undefined`... that's what `null` is for.

**Comment by Xerus2000** [comment]
Output finally working, but now every single HID controller is offset by one Byte, due to the `reportId` not being considered part of the payload anymore. Either we hack around that or we change all controller scripts. 
For the sake of not breaking custom scripts by others that are not yet integrated and since it is part of the actual data sent to the controller, I would propose changing it in the library, but I can also understand the other way.

**Comment by Be-ing** [comment]
Changing the output offsets in every existing HID script would be tedious and error prone, especially without having all of those controllers available to test with. So I'm leaning towards hacking the library and documenting that output offsets should start at 1, not 0. In a new library where we don't have to worry about breaking old scripts, we can change this.

**Comment by Be-ing** [comment]
Please make your commits as small as logically possible. When there is a lot of code cleanup in a commit, it is hard to find what functional changes are being made. In that case I think it's better to have a commit that's just for code cleanup. The changes to common-controller-scripts.js aren't relevant to this PR, so please revert them and you can open a new PR where we can focus on discussing those changes. Also, rebase onto the 2.2 branch so this fix can be included in 2.2.1.

File: res/controllers/common-hid-packet-parser.js [hunk 3]
```diff
@@ -652,26 +652,22 @@ HIDPacket.prototype.send = function() {
     }
 
     for (var group_name in this.groups) {
-        var group = this.groups[group_name];
+        const group = this.groups[group_name]
```
Comment (Be-ing) [inline res/controllers/common-hid-packet-parser.js]: While `const` somehow works without throwing any error in the old QtScript interpeter, I don't think it's actually `const`. To avoid confusion, please use `var` until we have ES7 available after merging #1795.

File: res/controllers/common-hid-packet-parser.js [hunk 4]
```diff
@@ -652,26 +652,22 @@ HIDPacket.prototype.send = function() {
...
         for (var field_name in group) {
-            var field = group[field_name];
-            if (field.type=="bitvector") {
```
Comment (Be-ing) [inline res/controllers/common-hid-packet-parser.js]: Are you sure it's safe to remove this? I haven't looked to ensure that no existing scripts use it. I think it's safest to make the smallest changes possible to get this working.

**Review by Be-ing — CHANGES_REQUESTED** [review]

File: .idea/codeStyles/Project.xml [hunk 5]
```diff
@@ -0,0 +1,29 @@
+<component name="ProjectCodeStyleConfiguration">
```
Comment (Be-ing) [inline .idea/codeStyles/Project.xml]: Looks like you accidentally committed some files

**Comment by Be-ing** [comment]
The order of the commits got mixed up somehow. Please do `git rebase -i` to rearrange them in the order they were actually done.

**Comment by Xerus2000** [comment]
Got an error on Linux:
```
Warning [Controller]: ControllerEngine: "Syntax error at line 663, column 13 in file /home/janek/.mixxx/controllers/common-hid-packet-parser.js: Expected `;', `;'"
Warning [Controller]: ControllerEngine: "Uncaught exception at line 9 in file /home/janek/.mixxx/controllers/Traktor-Kontrol-F1-scripts.js: ReferenceError: Can't find variable: HIDController\nBacktrace:\nKontrolF1Controller() at /home/janek/.mixxx/controllers/Traktor-Kontrol-F1-scripts.js:9\n<global>() at /home/janek/.mixxx/controllers/Traktor-Kontrol-F1-scripts.js:315"
Warning [Controller]: ControllerEngine: No "KontrolF1" object in script
```

File: res/controllers/common-hid-packet-parser.js [hunk 6]
```diff
@@ -640,37 +640,35 @@ HIDPacket.prototype.parse = function(data) {
...
-    //for (d in packet.data) {
-    //  if (packet.data[d] < 0x10) {
-    //    packet_string += "0";
-    //  }
-    //  packet_string += packet.data[d].toString(16) + " ";
-    //}
-    //HIDDebug("packet: " + packet_string);
-    controller.send(packet.data, packet.length, 0);
+    if(debug) {
+        let packet_string = "";
```
Comment (Be-ing) [inline res/controllers/common-hid-packet-parser.js]: `let` is not supported. Use `var`

File: res/controllers/common-hid-packet-parser.js (hunk 4 above)
Comment (Xerus2000) [reply res/controllers/common-hid-packet-parser.js]: No script _can_ use it. Please read fully what I have removed. It does nothing except repeatedly assigning `var bit` ^^

Comment (Be-ing) [reply res/controllers/common-hid-packet-parser.js]: Oh, I see. Well that's useless.

**Comment by Be-ing** [comment]
I propose subtracting 1 from the offsets passed to HIDPacket.prototype.addOutput. Alternatively you could subtract 1 from the field offset in HIDPacket.prototype.pack, but I think that would be harder to find for someone reading the code later.

**Review by uklotzde — CHANGES_REQUESTED** [review]
Please interactively rebase your commits to exclude the files in the `.idea` folder and force push the result afterwards. As @Be-ing already mentioned those files should not be commited into the main repository.

**Comment by Xerus2000** [comment]
Oh, didn't notice it. I know my history-editing fu and will fix everything up soon. But currently visiting family ^^

File: res/controllers/common-hid-packet-parser.js (hunk 3 above)
Comment (Xerus2000) [reply res/controllers/common-hid-packet-parser.js]: const is used a lot throughout the scripts though, just did a quick search. Should this all be var then?

Comment (Xerus2000) [reply res/controllers/common-hid-packet-parser.js]: ah this is all in the novation-launchpad thing - nevermind

**Comment by Xerus2000** [comment]
Something is wrong with GitHub here. Git is displaying the order correctly.

image

image

**Comment by Xerus2000** [comment]
Okay, this should be ready now, but I can't test it since I will only have controllers available again in a few weeks. So we either need to find other testers and wait. Also, I think I will take on rethinking this whole HID scripts situation, maybe develop a new script to parse packets for future controller scripts to be based on.

**Comment by Xerus2000** [comment]
@uklotzde @Be-ing Can we merge? Everything should work now.

**Review by uklotzde — APPROVED** [review]

**Comment by uklotzde** [comment]
Commit history is fixed :+1: I'm not able to asses the actual changes, though.

**Comment by Be-ing** [comment]
I'm not comfortable merging this before anyone has tested it with real HID controllers. We can merge it for 2.2.1.

**Comment by Xerus2000** [comment]
I can fully understand you. Don't you know others who use HID and could test this? Can we create groups on Zulip to ping for reviews or questions, e.g. a HID group?

**Review by Be-ing — CHANGES_REQUESTED** [review]
Please use 4 spaces to indent instead of tab characters. Some editors have an option to visually mark tab characters as separate from spaces to make this easier to see.

File: res/controllers/common-hid-packet-parser-test.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
+// Manual packing test functions
```
[the remark this answers is not in the document]
Comment (Xerus2000) [reply res/controllers/common-hid-packet-parser-test.js]: It was a bunch of code that was commented out in `common-hid-packet-parser.js` and seemed to be for testing. So I moved it out instead of deleting in case it was useful.

[the remark this answers is not in the document]
Comment (Be-ing) [reply res/controllers/common-hid-packet-parser-test.js]: As far as I can tell, this doesn't really test anything, it's just a tool for debugging. I'd prefer to move it back to common-hid-packet-parser.js than have this stray file.

[the remark this answers is not in the document]
Comment (Xerus2000) [reply res/controllers/common-hid-packet-parser-test.js]: I don't like to have this stray piece of code in a file I regularly use as a reference.

**Comment by Be-ing** [comment]
Have you had a chance to test this on Windows with an HID controller yet?

File: res/controllers/common-hid-packet-parser-test.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (Be-ing) [reply res/controllers/common-hid-packet-parser-test.js]: Maybe we should just delete it?

[the remark this answers is not in the document]
Comment (Xerus2000) [reply res/controllers/common-hid-packet-parser-test.js]: Yes. It will still be in the history if somebody needs it ever again

**Comment by Xerus2000** [comment]
I just returned home, so I can check it in the next few days.

**Comment by Xerus2000** [comment]
Okay, I am on Linux right now, and have some really odd behaviour:
I have symlinked the controllers folder `/usr/share/mixxx/controllers` to point at the controllers folder of my local mixxx clone. I was on the master, the 2.2 and the kontrolz1 branch, all worked fine. Then I checked out this branch, and all of a sudden, all my controllers
- took multiple seconds to load
- showed only little or no light output
- mixxx responded with extrem delay and lag to changes to the controllers controls

I have experienced this twice already, and I don't know what the heck is happening here. The first time it somehow went away after some toying around, but even then, it didn't go away by checking out one of the other branches, and it isn't going away now. Apparently, checking out this branch leaves some weird traces in Mixxx... ? I don't even know what to think, this doesn't make sense...

**Comment by Xerus2000** [comment]
I have just restarted my computer, reinstalled mixxx and deleted `~/.mixxx`, but this problem just persists. What the heck is going on here? I will go on Windows and see if I experience similar issues...

**Comment by Xerus2000** [comment]
Even more fun on Windows: Mixxx does not shut down when i click close, even when i wait for a few minutes. This is gonna be fun...

**Comment by Pegasus-RPG** [comment]
@Xerus Do you have a failing hard drive?

**Comment by Xerus2000** [comment]
No

**Comment by Xerus2000** [comment]
Okay, so somehow, the lights now work completely fine on Windows, but the input is just simply ignored. Or to be specific, it is reported by controllerDebug, but the script is not getting it.

**Comment by Be-ing** [comment]
I symlink ~/.mixxx/controllers to my Git repository as [described on the wiki](https://mixxx.org/wiki/doku.php/using_git#working_on_mappings_and_skins_separately_from_other_changes) rather than my system controllers folder.

**Comment by Xerus2000** [comment]
Hah, but that one is linked to Dropbox to sync my custom mappings. But it's an interesting hint. 
The scripts in the user folder always take precedence over the global folder, right? I was quite confused by Mixxx always copying "common-hid-packet-parser.js" to the user folder.

Also, once again the problem sits in front of the computer. I didn't adapt the packet definitions of my mappings to the PR -.- I'll have look into this another time, there seem to be some unresolved issues.

**Comment by Xerus2000** [comment]
Nope, the controller mappings weren't the thing. I now have a script to swap out both controllers folders instantly between the testing one and the stable one. But whenever I switch to the controller scripts of this PR, the lag stays even after swapping back. I'm calling Voodoo magic here! I am only changing the controllers folders back and forth, and mixxx somehow retains errors from one configuration. Is it caching the scripts somehow? This is crazy!
And then, a few days later, the problem somehow magically vanishes. This is _VERY_ weird.

**Comment by Xerus2000** [comment]
Aha, I have at least found a solution to this problem: Unplugging the controllers and then plugging them back in and restarting mixxx. Maybe we have unintentionally sent some special signals to the controllers :thinking:

**Comment by Xerus2000** [comment]
I have found the root of the issue! FINALLY! But at first sight the issue looks like a fundamental incompability of Linux and Windows...

**Comment by Xerus2000** [comment]
Alright, figured it out. This at first seemingly simple review ate so many hours of my live -.-
Anyways, here's the deal:
At least for all my controllers, the last parameter of `controller.send` MUST be `0` in Linux and `this.reportId` in Windows. I have tested back and forth on both, and this is the only way that it works.

Btw, the old Kontrol F1 script for some reason used `packet.addControl` instead of `packet.addOutput` for adding output, and I copied that in my script, thus the offset fix didn't work. I have fixed it for all my mappings now, but that is something to be careful about.

**Comment by Be-ing** [comment]
:confused: I'm skeptical there isn't a way to have one script working on both Windows and Linux. If so, that would defeat the point of hidapi. Have you verified by running Mixxx with `--controllerDebug` that you are sending the same data on both OSs?

**Comment by Xerus2000** [comment]
Okay, apparently the Z1 actually works fine with the windows configuration on Linux. But the F1 will receive no output at all and the input is super slow if the reportId is included.

This is a line of the log:
```
Warning [Controller]: Unable to send data to "Traktor Kontrol F1 0A3A_0" serial # "B1B50A3A" : ""
```

**Comment by Xerus2000** [comment]
Ok, I retract the previous. It all works now, with the same code on Windows and Linux, I have rebooted multiple times and there are no issues anymore. I don't know what happened there before ^^

File: res/controllers/common-hid-packet-parser.js [hunk 8]
```diff
@@ -1023,18 +1027,26 @@ HIDController.prototype.parsePacket = function(data,length) {
...
+        if (packet.reportId !== 0 && packet.reportId !== data[0]) {
             continue;
         }
-        // Check for packet header match against data
-        for (var header_byte=0;header_byte<packet.header.length;header_byte++) {
-            if (packet.header[header_byte]!=data[header_byte]) {
-                packet=undefined;
-                break;
+
+        if(packet.header !== undefined) {
```
Comment (Be-ing) [inline res/controllers/common-hid-packet-parser.js]: space after `if`

File: res/controllers/common-controller-scripts.js [hunk 9]
```diff
@@ -16,35 +14,65 @@ String.prototype.toInt = function() {
 // ----------------- Function overloads ---------------------
 
 // Causes script print() calls to appear in the log file as well
-print = function(string) {
+function print(string) {
```
Comment (Be-ing) [inline res/controllers/common-controller-scripts.js]: I do not like this function declaration syntax in JS. It obscures that all functions are objects, which can be quite confusing for developers not very familiar with JS. Please revert the changes to this file that are not related to this PR.

**Review by Be-ing — CHANGES_REQUESTED** [review]
Awesome, I'm glad we got this working! Just a few nitpicks before merging this.

File: res/controllers/common-hid-packet-parser.js [hunk 10]
```diff
@@ -1,28 +1,28 @@
 
-// Common HID script debugging function. Just to get logging with 'HID' prefix.
+/** Common HID script debugging function. Just to get logging with 'HID' prefix. */
```
[the remark this answers is not in the document]
Comment (Xerus2000) [reply res/controllers/common-hid-packet-parser.js]: Because then the IDE interprets it as documentation and shows it on hovering over the method.

File: res/controllers/common-hid-packet-parser.js [hunk 11]
```diff
@@ -610,6 +613,7 @@ HIDPacket.prototype.parse = function(data) {
                     field.value = value;
                     continue;
                 }
+                var change
```
[the remark this answers is not in the document]
Comment (Xerus2000) [reply res/controllers/common-hid-packet-parser.js]: I'd rather remove all semicolons than add one ^^ but ok

File: res/controllers/common-hid-packet-parser.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Be-ing) [reply res/controllers/common-hid-packet-parser.js]: That's fine, I was just wondering.

File: res/controllers/common-hid-packet-parser.js [hunk 12]
```diff
@@ -636,80 +640,80 @@ HIDPacket.prototype.parse = function(data) {
...
-    //  packet_string += packet.data[d].toString(16) + " ";
-    //}
-    //HIDDebug("packet: " + packet_string);
-    controller.send(packet.data, packet.length, 0);
+    if(debug) {
+        var packet_string = "";
+        for (var d in data) {
+            if (data[d] < 0x10) {
+                // Add padding for bytes smaller than 10
+                packet_string += "0";
```
[the remark this answers is not in the document]
Comment (Xerus2000) [reply res/controllers/common-hid-packet-parser.js]: ummm... No?

[the remark this answers is not in the document]
Comment (Xerus2000) [reply res/controllers/common-hid-packet-parser.js]: Grand news: DJ Software Mixxx introduces ONLINE PADDING! WOOHOO!

[the remark this answers is not in the document]
Comment (Be-ing) [reply res/controllers/common-hid-packet-parser.js]: It was a joke.

File: res/controllers/common-controller-scripts.js (hunk 9 above)
Comment (Xerus2000) [reply res/controllers/common-controller-scripts.js]: but it makes much more sense for people coming from other languages? They don't have to consider that functions are objects, they just want to call them and the syntax I introduced is much more familiar when looking at the code from that perspective.

**Comment by Xerus2000** [comment]
I would also appreciate if you could find someone with HID controllers to test this on any operating system, just to be extra-sure.

**Comment by Be-ing** [comment]
I agree, we should try to find someone to test this with other controllers before releasing 2.1.1. Someone who reported a few bugs on Launchpad recently mentioned they use a Kontrol S4 Mk2.

File: res/controllers/common-hid-packet-parser.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (Be-ing) [reply res/controllers/common-hid-packet-parser.js]: If you were starting a new file, sure. But let's keep the style consistent within this already existing file.

File: res/controllers/common-controller-scripts.js (hunk 9 above)
Comment (Be-ing) [reply res/controllers/common-controller-scripts.js]: It makes sense until they come across methods being called on functions and variables assigned to functions.

**Comment by Xerus2000** [comment]
Could you add the 2.2.1 Milestone to this? I definitely want this to be included, otherwise I have to maintain two parallel versions of my HID scripts...

**Comment by Be-ing** [comment]
Could you fix the minor issues noted above? Then we can merge this and you won't have to worry about that anymore.

File: res/controllers/common-hid-packet-parser.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (Xerus2000) [reply res/controllers/common-hid-packet-parser.js]: Well... I could do a find-replace and remove em all ^^

**Comment by Xerus2000** [comment]
Alright, after looking at the code I have tried to choose a reasonable code style and reformat the files according to it. I will test it again this evening, just to be sure. 

If you don't agree with choices in the codestyle, just let me know. We should have some common definition for javascript codestyle tbh, because right now it is naturally all over a place and with a proper IDE you can simply reformat it correctly.

File: res/controllers/common-controller-scripts.js [hunk 13]
```diff
@@ -14,15 +14,15 @@ String.prototype.toInt = function() {
...
     engine.log(string);
 }
 
-function printObject(obj, maxdepth) {
+printObject = function(obj, maxdepth) {
     print(stringifyObject(obj, maxdepth));
 }
 
-function stringifyObject(obj, maxdepth, checked, prefix) {
+stringifyObject = function(obj, maxdepth, checked, prefix) {
```
Comment (Be-ing) [inline res/controllers/common-controller-scripts.js]: Please remove these new functions. They use new JS features that will not be supported before #1795 is merged and are off topic from this PR.

**Review by Be-ing — CHANGES_REQUESTED** [review]

**Comment by Be-ing** [comment]
We have code style guidelines on the [wiki](https://mixxx.org/wiki/doku.php/contributing_mappings#coding_conventions_for_javascript) although we haven't been enforcing them very strictly. If you want to talk about that further, please start a new topic on Zulip. Generally stable branches are not a great place to be making large code changes, even if they are just formatting.

File: res/controllers/common-controller-scripts.js (hunk 13 above)
Comment (Be-ing) [reply res/controllers/common-controller-scripts.js]: If they happen to work it is just by coincidence that old JS interpreter tolerates them, but they may not really be working as you expect if you give them different input.

**Comment by Be-ing** [comment]
> I will test it again this evening, just to be sure.

Yes, let's make sure to test with real hardware before merging. After merging this, let's try to find people to test with other HID controllers on Windows before releasing 2.2.1.

File: res/controllers/common-controller-scripts.js [hunk 14]
```diff
@@ -47,7 +47,7 @@ function stringifyObject(obj, maxdepth, checked, prefix) {
     return obj;
 }
 
-function arrayContains(array, elem) {
+arrayContains = function(array, elem) {
```
Comment (Be-ing) [inline res/controllers/common-controller-scripts.js]: ES7 has a standard function for this: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes

Comment (Xerus2000) [reply res/controllers/common-controller-scripts.js]: I somehow didn't get that to work, I thought we are using some old JS version?

File: res/controllers/common-controller-scripts.js (hunk 13 above)
Comment (Xerus2000) [reply res/controllers/common-controller-scripts.js]: Just curious, which JS features are you referring to? The function has served me well on many different inputs.

File: res/controllers/common-controller-scripts.js (hunk 14 above)
Comment (Be-ing) [reply res/controllers/common-controller-scripts.js]: Correct, we are. I mentioned that to point out that this function will be redundant with the language built in function once #1795 is merged, so I don't think we should add it now.

**Comment by Xerus2000** [comment]
I have tested it, at least for my controllers everything works fine. It could be that other controllers scripts have similar issues to mine if they use `addControl` instead of `addOutput` in the outgoing packets, but that's easily fixable and just concerns the offset.

File: res/controllers/common-controller-scripts.js (hunk 13 above)
Comment (Be-ing) [reply res/controllers/common-controller-scripts.js]: My bad, I thought `try...catch` was a newer JS feature, but that has been around a long time. `const`, however, is not implemented in the current JS interpeter.

Comment (Xerus2000) [reply res/controllers/common-controller-scripts.js]: Then I'll just replace const by var and done ^^

File: res/controllers/common-controller-scripts.js (hunk 14 above)
Comment (Xerus2000) [reply res/controllers/common-controller-scripts.js]: Well, it is needed for toString in the meanwhile. It's not like it will create any conflicts. But it's a good point, we'll remove it when that PR is merged.

**Comment by Xerus2000** [comment]
Anyways, I think we should merge this now and then take care of the code style separately.

**Comment by Xerus2000** [comment]
Is anything missing?

**Comment by Be-ing** [comment]
We can make a note about that when posting about the 2.2.1 release in the forum.

> people who have custom HID mappings and are not in contact with us

I do not assume any such people exist considering this serious bug went unnoticed for 5 years.

**Comment by Xerus2000** [comment]
Well, I guess people who use Native Instruments and are not on Linux will most likely use Traktor Pro, so that combination Windows + HID + Mixxx is rather unlikely considering that already rules out a few HID devices. 
Anyways, I guess you are right. People who go through the trouble of mapping HID will probably sooner or later come and ask about something and contribute the result.

If we merge this now, it will be in the next 2.2 launchpad build, right? Then I could finally stop maintaining the duplicate scripts... So please merge :)

**Comment by Be-ing** [comment]
Launchpad only makes builds for the PPA. Our own Jenkins servers make the Windows and macOS builds, which are uploaded to http://downloads.mixxx.org/builds/

**Comment by Xerus2000** [comment]
Wait, this is now merged into 2.2 but not in master. How do you handle your version branching?

**Comment by daschuer** [comment]
All 2.1 and 2.2 changes have been merge to master as well.
